src/loki_cluster.py

```diff
@@ -0,0 +1,317 @@
+
+
+def expand_roles(roles: Iterable[LokiRole]) -> Set[LokiRole]:
+    """Expand any meta roles to their 'atomic' equivalents."""
+    expanded_roles = set()
+    for role in roles:
+        if role in META_ROLES:
+            expanded_roles.update(META_ROLES[role])
+        else:
+            expanded_roles.add(role)
+    return expanded_roles
+
+
+class ConfigReceivedEvent(ops.EventBase):
+    """Event emitted when the "loki-cluster" provider has shared a new loki config."""
+
+    config: Dict[str, Any]
+    """The loki config."""
+
+    def __init__(self, handle: ops.framework.Handle, config: Dict[str, Any]):
+        super().__init__(handle)
+        self.config = config
+
+    def snapshot(self) -> Dict[str, Any]:
+        """Used by the framework to serialize the event to disk.
+
+        Not meant to be called by charm code.
+        """
+        return {"config": json.dumps(self.config)}
+
+    def restore(self, snapshot: Dict[str, Any]):
+        """Used by the framework to deserialize the event from disk.
+
+        Not meant to be called by charm code.
+        """
+        self.relation = json.loads(snapshot["config"])
+
+
+class LokiClusterError(Exception):
+    """Base class for exceptions raised by this module."""
+
+
+class DataValidationError(LokiClusterError):
+    """Raised when relation databag validation fails."""
+
+
+class DatabagAccessPermissionError(LokiClusterError):
+    """Raised when a follower attempts to write leader settings."""
+
+
+class JujuTopology(pydantic.BaseModel):
+    """JujuTopology."""
+
+    model: str
+    unit: str
+    # ...
+
+
+class DatabagModel(BaseModel):
+    """Base databag model."""
+
+    model_config = ConfigDict(
+        # Allow instantiating this class by field name (instead of forcing alias).
+        populate_by_name=True,
+        # Custom config key: whether to nest the whole datastructure (as json)
+        # under a field or spread it out at the toplevel.
+        _NEST_UNDER=None,
+    )  # type: ignore
+    """Pydantic config."""
+
+    @classmethod
+    def load(cls, databag: MutableMapping[str, str]):
+        """Load this model from a Juju databag."""
+        nest_under = cls.model_config.get("_NEST_UNDER")
+        if nest_under:
+            return cls.parse_obj(json.loads(databag[nest_under]))
+
+        try:
+            data = {k: json.loads(v) for k, v in databag.items() if k not in BUILTIN_JUJU_KEYS}
+        except json.JSONDecodeError as e:
+            msg = f"invalid databag contents: expecting json. {databag}"
+            log.info(msg)
+            raise DataValidationError(msg) from e
+
+        try:
+            return cls.parse_raw(json.dumps(data))  # type: ignore
+        except pydantic.ValidationError as e:
+            msg = f"failed to validate databag: {databag}"
+            log.info(msg, exc_info=True)
+            raise DataValidationError(msg) from e
+
+    def dump(self, databag: Optional[MutableMapping[str, str]] = None, clear: bool = True):
```

Comment:
I think we should decouple between the basemodel class and ops.
I.e. instead of `databag.dump(relation.data)` we could do `dump(databag, relation)`.

This may obviate the need for our own custom `dump` method.

Reply:
Yes, we can do that... anyway it not seems unreasonable to have that method in the Pydantic model.

Are you thinking in something like this?

```python
def dump(model: DatabagModel, databag: Optional[MutableMapping[str, str]] = None, clear: bool = True):
    """Write the contents of the model to Juju relation databag.

    :param model: the pydantic model
    :param databag: the databag to write the data to.
    :param clear: ensure the databag is cleared before writing it.
    """
    if clear and databag:
        databag.clear()

    if databag is None:
        databag = {}
        nest_under = model.model_config.get("_NEST_UNDER")
    if nest_under:
        databag[nest_under] = model.json()

    dct = model.model_dump(by_alias=True)
    for key, field in model.model_fields.items():  # type: ignore
        value = dct[key]
        databag[field.alias or key] = json.dumps(value)
    return databag
```

Reply:
Kind of, just without `NEST_UNDER` and without `clear`.

Reply:
https://github.com/canonical/loki-worker-k8s-operator/issues/6